rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,260 @@
+
+## Scope
+
+This RFC will cover:
+
+- How and where the pipelines are stored
+- How vector reads the pipelines and what are the limits of pipelines
+- What metadatas vector adds to the events going through a pipeline
+
+This RFC will not cover:
+
+- How the pipelines should be synchronized between vector instances
+- Pre/post processing of data - The ability to prepare and normalize data.
+- Connecting pipelines together - The ability to take input from another pipeline.
+- Component reuse - The ability to define boilerplate for reuse across many different pipelines. This will likely align with Datadog’s “pipeline catalogue”.
+- Pipeline quotas - The ability to limit how much data a pipeline can send to a sink.
+
+
+## Motivation
+
+- Helps Vector to grow organically within an organization by allowing teams to adopt Vector at their own pace without heavy administrator involvement.
+- Reduces the management overhead of devops/SREs by enabling teams to manage their own pipelines (spread the management load).
+
+
+## Internal Proposal
+
+### How and where the pipelines are stored
+
+To keep the actual structure and avoid retro compatibility issues, vector should read from the default folder `/etc/vector/pipelines/*` the difference pipeline's configuration files.
+The pipeline directory should have a flat structure (no sub directories) and only `json`, `toml` and `yaml` files will be supported.
+
+To discuss: make this folder should be configurable through the cli, the same way than for the configuration file, with `--pipelines-dir` or though and environment variable `VECTOR_PIPELINES_DIR`.
+
+### How vector reads the pipelines and what are the limits
+
+Vector will read each of the pipeline configuration files at start time and create an `id` attribute corresponding to the pipeline's filename. A `load-balancer.yml` will have `load-balancer` as `id`. In the pipeline's configuration file, the `id` can be set manually.
+In addition, a `version` attribute will be set containing the hash of the pipeline's configuration file to keep track of the file changes.
+In case several files having the same `id` in that folder (for example `load-balancer.yml` and `load-balancer.json`), vector should error.
+
+Those pipeline's configuration files should only contain transforms or vector will error.
+Each of the defined transforms will be callable by using `pipeline-id.transform-id` outside of the pipeline's configuration file to avoid conflicts.
+A pipeline cannot use a component defined in another pipeline of vector will error.
+
+The structure of the pipeline's configuration file should be as follow
+
+```toml
+id = "load-balancer"
```

Comment:
If we require this then the file structure becomes required and and we remove the ability to have multiple pipelines in a single file. Would a more flexible approach be requiring that each component is namespaced to a specific pipeline like:

```
pipelines.load_balancer.transforms.first
...
```